### ISSUE TITLE
Easier optic composition

### ISSUE THREAD

**Issue by maackle** [TOETOE55/lens-rs#13]
If I want to define some optics and then compose them later, I can't use the `optics!` macro, because it binds the optics chain to the identity `__`. It could be nice to have `optics!` return the type constructor rather than a fully completed optic, so that optics can be composed.

For instance if I want to construct an optic out of some other optics I constructed, I can do this:

```rust
let x = todo!("something complex");

let full = optics!(_1.Left._1);

let a = |o| optics::Left(optics::_1(o));
let b = |o| optics::_1(o);
let composed = b(a(optics::__));

assert_eq!(x.preview_ref(full), x.preview_ref(composed));
```

`a` and `b` are useful optics in their own right -- they could be used to as building blocks for other optics. But I have to use the usual rust syntax for this, and the closures are not recognized as lenses by `lens_rs`. I wonder if optics could be redefined to be functions rather than structs so that they can be endlessly composed. Then `view`, `preview`, etc could take functions and supply `__` itself.

Perhaps `optics!` could have syntax for this too. But even just allowing function composition could be good enough.

**Comment by TOETOE55** [comment]
Did you means that `optics!(a.{b})` is directly generated as  `|o| optics::a(b(o))`(intro a new syntax `{b}` to capture constructed optic), and `View`, `Preview`, etc take the function `Fn(__) -> Optic` instead of `Optic`?🤔

I may be more worried about the closure form of `optics!` makes it more difficult for type inference, and it makes the traits harder to understand, and need more effort to make them dyn safe.

**Comment by maackle** [comment]
Yes, exactly -- or something like that. And yes, it may be difficult to line up the types for type inference. And yes, it would be a bit harder to understand, if it even works.

I wonder if there are other ways to make lenses composable though. That is a key feature of Haskell lenses and van Laarhoven lenses in general, that they compose very well.

**Comment by TOETOE55** [comment]
In fact, the optic designed at the beginning looked like this in order to better compose optics

```rust
struct x;
struct y;

struct Compose<O1, O2>(O1, O2);

impl Lens<x> for Struct { 
    type Image;
    /*...*/ 
}

impl<O1, O2, Y> Lens<Compose<O1, O2>> for Y
where
    Lens<O1>::Image: Lens<O2>,
    Y: Lens<O1>
{
    type Image = Lens<O2>::Image;
    /*...*/
}

```

But this design makes it impossible to infer the type of some situations and makes some optics inexpressible, I gave up.

> Did you means that optics!(a.{b}) is directly generated as |o| optics::a(b(o))(intro a new syntax {b} to capture constructed optic), and View, Preview, etc take the function Fn(__) -> Optic instead of Optic?🤔

I haven't experimented with this idea yet, I think I can try to do it.

(Although I think the current combination is not bad)

**Comment by TOETOE55** [comment]
I'd like some apis that `.a().b().c()` returning `optics::a(optics::b(optics::c(__)))` instead of macro. But I have no idea to do it yet.
